src/test/SecurityPanel/StoreJmxCredentials.test.tsx

```diff
@@ -0,0 +1,259 @@
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+import * as React from 'react';
+import userEvent from '@testing-library/user-event';
+import renderer, { act } from 'react-test-renderer';
+import { render, screen } from '@testing-library/react';
+import { of } from 'rxjs';
+
+import '@testing-library/jest-dom';
+import { StoreJmxCredentials } from '@app/SecurityPanel/StoreJmxCredentials';
+import { ServiceContext, defaultServices } from '@app/Shared/Services/Services';
+import { Modal, ModalVariant } from '@patternfly/react-core';
+import { NotificationMessage } from '@app/Shared/Services/NotificationChannel.service';
+import { Target } from '@app/Shared/Services/Target.service';
+
+jest.mock('@app/SecurityPanel/CreateJmxCredentialModal', () => {
+  return {
+    CreateJmxCredentialModal: jest.fn((props) => {
+      return (
+        <Modal
+          isOpen={props.visible}
+          variant={ModalVariant.large}
+          showClose={true}
+          onClose={props.onClose}
+          title="CreateJmxCredentialModal"
+        >
+          Jmx Auth Form
+        </Modal>
+      );
+    }),
+  };
+});
+
+jest.mock('@app/Shared/Services/Api.service', () => {
+  const MOCK_TARGET = { connectUrl: 'service:jmx:rmi://someUrl', alias: 'fooTarget' } as Target;
```

Comment:
Technically, I should be able to prefix the variables with `mock` to have them accessible inside `jest.mock()` [(jest docs)](https://jestjs.io/docs/es6-class-mocks#calling-jestmock-with-the-module-factory-parameter), but I'm still getting the same error as above. Also tried variations of casting `as Target`, `as Target[]` and using the spread operator `.mockReturnValueOnce(of([{...mockTarget} as Target])`.
```
const mockTarget = { connectUrl: 'service:jmx:rmi://someUrl', alias: 'fooTarget' } as Target;
const mockAnotherTarget = { connectUrl: 'service:jmx:rmi://anotherUrl', alias: 'anotherTarget' } as Target;
...
ApiService: jest.fn(() => {
      return {
        getTargetsWithStoredJmxCredentials: jest
          .fn()
          .mockReturnValueOnce(of([mockTarget as Target]))
...
```

Any suggestions?

Reply:
Since the issue is caused by `jest.mock()` calls being hoisted to the top of the file above any variable declarations, then [`jest.doMock`](https://jestjs.io/docs/jest-object#jestdomockmodulename-factory-options) is another option. Since `doMock` isn't hoisted to the top of the file, theoretically any out-of-scope variables should be defined. I haven't used this function yet though but let me know if you need help.

Reply:
After trying several [workarounds](https://github.com/facebook/jest/issues/2567) using `jest.mock`, I've been working on getting the tests to run with `jest.doMock`. The docs you linked seem to duplicate the `jest.doMock` and ` const moduleName = require('../moduleName');` for every test though. 

@hareetd Do you know if putting all of my `jest.doMock()`s inside a `beforeAll` will result in the same behaviour as the docs, but without duplicating code?

I've committed what I have so far - right now the test compilation fails with `Error: Uncaught [TypeError: Cannot read properties of undefined (reading 'subscribe')]`. Seems like Jest can't find the mocked service functions?

Reply:
Hey, it looks like putting `jest.doMock` or even `jest.mock` inside a `beforeAll`/`beforeEach` should [work](https://github.com/facebook/jest/issues/2582#issuecomment-284514827). I think the issue is that since the `ApiService` and `TargetsService` mocks are now being defined in the `describe` block, they show up as undefined when the `StoreJmxCredentials` import happens on line 46. 

I wonder if importing `StoreJmxCredentials` inside the `beforeAll` block would fix it; might be worth experimenting with.

Reply:
Thanks, fixed it by moving the `doMock()`s outside of the `describe` block, then adding the `StoreJmxCredentials`, `ServiceContext`, and `defaultServices` imports after the `doMock()`s and before `describe`. I should add a note in the readme for future reference.

Reply:
Nice, glad it's working! Yeah, this is definitely a good thing to note in the `TESTING.md` file.